File: python/bc-ur/src/bc_ur/_fountain_part.py
"""Fountain code part with manual CBOR serialization.

Fountain parts are encoded as CBOR arrays matching the minicbor format:
  [sequence, sequence_count, message_length, checksum, data]

We use manual CBOR construction (not dcbor) to produce byte-identical output
to minicbor, which uses standard (non-deterministic) CBOR encoding.
"""

from __future__ import annotations

from ._fountain_utils import choose_fragments
from .error import FountainError


def _encode_cbor_unsigned(value: int) -> bytes:
    """Encode an unsigned integer in CBOR (major type 0)."""
    if value < 0:
        raise FountainError("CBOR unsigned integer must be non-negative")
    if value <= 23:
        return bytes([value])
    elif value <= 0xFF:
        return bytes([0x18, value])
    elif value <= 0xFFFF:
        return bytes([0x19]) + value.to_bytes(2, "big")
    elif value <= 0xFFFFFFFF:
        return bytes([0x1A]) + value.to_bytes(4, "big")
    else:
        return bytes([0x1B]) + value.to_bytes(8, "big")


def _encode_cbor_bytes(data: bytes) -> bytes:
    """Encode a byte string in CBOR (major type 2)."""
    length = len(data)
    if length <= 23:
        return bytes([0x40 + length]) + data
    elif length <= 0xFF:
        return bytes([0x58, length]) + data
    elif length <= 0xFFFF:
        return bytes([0x59]) + length.to_bytes(2, "big") + data
    else:
        return bytes([0x5A]) + length.to_bytes(4, "big") + data


def _decode_cbor_unsigned(data: memoryview, pos: int) -> tuple[int, int]:
    """Decode a CBOR unsigned integer, returning (value, new_pos)."""
    first = data[pos]
    major = first >> 5
    info = first & 0x1F
    if major != 0:
        raise FountainError(f"expected unsigned integer at position {pos}")
    if info <= 23:
        return info, pos + 1
    elif info == 24:
        return data[pos + 1], pos + 2
    elif info == 25:
        return int.from_bytes(data[pos + 1 : pos + 3], "big"), pos + 3
    elif info == 26:
        return int.from_bytes(data[pos + 1 : pos + 5], "big"), pos + 5
    elif info == 27:
        val = int.from_bytes(data[pos + 1 : pos + 9], "big")
        if val > 0xFFFFFFFF:
            raise FountainError("value exceeds u32 range")
        return val, pos + 9
    else:
        raise FountainError(f"unsupported CBOR additional info {info}")


def _decode_cbor_bytes(data: memoryview, pos: int) -> tuple[bytes, int]:
    """Decode a CBOR byte string, returning (bytes, new_pos)."""
    first = data[pos]
    major = first >> 5
    info = first & 0x1F
    if major != 2:
        raise FountainError(f"expected byte string at position {pos}")
    if info <= 23:
        length = info
        start = pos + 1
    elif info == 24:
        length = data[pos + 1]
        start = pos + 2
    elif info == 25:
        length = int.from_bytes(data[pos + 1 : pos + 3], "big")
        start = pos + 3
    elif info == 26:
        length = int.from_bytes(data[pos + 1 : pos + 5], "big")
        start = pos + 5
    else:
        raise FountainError(f"unsupported byte string length encoding at position {pos}")
    return bytes(data[start : start + length]), start + length


class FountainPart:
    """A part emitted by a fountain encoder."""

    __slots__ = ("_sequence", "_sequence_count", "_message_length", "_checksum", "_data")

    def __init__(
        self,
        sequence: int,
        sequence_count: int,
        message_length: int,
        checksum: int,
        data: bytes,
    ) -> None:
        self._sequence = sequence
        self._sequence_count = sequence_count
        self._message_length = message_length
        self._checksum = checksum
        self._data = data

    @property
    def sequence(self) -> int:
        return self._sequence

    @property
    def sequence_count(self) -> int:
        return self._sequence_count

    @property
    def message_length(self) -> int:
        return self._message_length

    @property
    def checksum(self) -> int:
        return self._checksum

    @property
    def data(self) -> bytes:
        return self._data

    def indexes(self) -> list[int]:
        """Return the fragment indexes combined into this part."""
        return choose_fragments(self._sequence, self._sequence_count, self._checksum)

    def is_simple(self) -> bool:
        """Whether this part represents a single original fragment."""
        return len(self.indexes()) == 1

    def sequence_id(self) -> str:
        """Return the sequence identifier string (e.g., '1-9')."""
        return f"{self._sequence}-{self._sequence_count}"

    def to_cbor(self) -> bytes:
        """Encode this part as CBOR bytes (matching minicbor format)."""
        parts = [
            bytes([0x85]),  # array of 5
            _encode_cbor_unsigned(self._sequence),
            _encode_cbor_unsigned(self._sequence_count),
            _encode_cbor_unsigned(self._message_length),
            _encode_cbor_unsigned(self._checksum),
            _encode_cbor_bytes(self._data),
        ]
        return b"".join(parts)

    @classmethod
    def from_cbor(cls, cbor_data: bytes) -> FountainPart:
        """Decode a FountainPart from CBOR bytes."""
        data = memoryview(cbor_data)
        pos = 0

        # Expect array of 5 (CBOR header byte 0x85)
        if data[pos] != 0x85:
            raise FountainError(
                f"expected CBOR array of 5 elements, got 0x{data[pos]:02x}"
            )
        pos += 1

        sequence, pos = _decode_cbor_unsigned(data, pos)
        sequence_count, pos = _decode_cbor_unsigned(data, pos)
        message_length, pos = _decode_cbor_unsigned(data, pos)
        checksum, pos = _decode_cbor_unsigned(data, pos)
        part_data, pos = _decode_cbor_bytes(data, pos)

        return cls(sequence, sequence_count, message_length, checksum, part_data)

    def __eq__(self, other: object) -> bool:
        if isinstance(other, FountainPart):
            return (
                self._sequence == other._sequence
                and self._sequence_count == other._sequence_count
                and self._message_length == other._message_length
                and self._checksum == other._checksum
                and self._data == other._data
            )
        return NotImplemented

    def __repr__(self) -> str:
        return (
            f"FountainPart(seq={self._sequence}, count={self._sequence_count}, "
            f"msglen={self._message_length}, checksum=0x{self._checksum:08x}, "
            f"datalen={len(self._data)})"
        )

File: go/bcur/fountain_part.go
package bcur

import (
	"encoding/binary"
	"fmt"
)

// fountainPart represents a single part in fountain-coded communication.
type fountainPart struct {
	sequence      int
	sequenceCount int
	messageLength int
	checksum      uint32
	data          []byte
}

// indexes returns the fragment indices combined in this part.
func (p *fountainPart) indexes() []int {
	return chooseFragments(p.sequence, p.sequenceCount, p.checksum)
}

// isSimple returns true if this part represents a single original fragment.
func (p *fountainPart) isSimple() bool {
	return len(p.indexes()) == 1
}

// sequenceID returns a string like "1-9" identifying this part.
func (p *fountainPart) sequenceID() string {
	return fmt.Sprintf("%d-%d", p.sequence, p.sequenceCount)
}

// clone returns a deep copy of this part.
func (p *fountainPart) clone() *fountainPart {
	dataCopy := make([]byte, len(p.data))
	copy(dataCopy, p.data)
	return &fountainPart{
		sequence:      p.sequence,
		sequenceCount: p.sequenceCount,
		messageLength: p.messageLength,
		checksum:      p.checksum,
		data:          dataCopy,
	}
}

// toCBOR encodes this part as a CBOR byte array. Uses manual CBOR encoding
// to match minicbor's non-deterministic output format.
func (p *fountainPart) toCBOR() []byte {
	var buf []byte
	// Array of 5 items
	buf = append(buf, 0x85)
	buf = appendCBORUnsigned(buf, uint64(p.sequence))
	buf = appendCBORUnsigned(buf, uint64(p.sequenceCount))
	buf = appendCBORUnsigned(buf, uint64(p.messageLength))
	buf = appendCBORUnsigned(buf, uint64(p.checksum))
	buf = appendCBORBytes(buf, p.data)
	return buf
}

// fountainPartFromCBOR decodes a CBOR-encoded fountain part.
func fountainPartFromCBOR(data []byte) (*fountainPart, error) {
	if len(data) == 0 {
		return nil, ErrCBORDecode
	}

	offset := 0

	// Must be array of 5
	if offset >= len(data) {
		return nil, ErrCBORDecode
	}
	majorType := data[offset] >> 5
	if majorType != 4 { // major type 4 = array
		return nil, ErrCBORDecode
	}
	arrayLen, n, err := decodeCBORUint64(data, offset)
	if err != nil {
		return nil, err
	}
	offset = n
	if arrayLen != 5 {
		return nil, ErrCBORArrayLength
	}

	// Decode 4 unsigned integers
	values := make([]uint32, 4)
	for i := 0; i < 4; i++ {
		v, n, err := decodeCBORUint(data, offset)
		if err != nil {
			return nil, err
		}
		offset = n
		if v > 0xFFFFFFFF {
			return nil, fmt.Errorf("value too large for u32 at position %d", offset)
		}
		values[i] = uint32(v)
	}

	// Decode byte string
	partData, _, err := decodeCBORByteString(data, offset)
	if err != nil {
		return nil, err
	}

	return &fountainPart{
		sequence:      int(values[0]),
		sequenceCount: int(values[1]),
		messageLength: int(values[2]),
		checksum:      values[3],
		data:          partData,
	}, nil
}

// appendCBORUnsigned encodes an unsigned integer in CBOR format.
func appendCBORUnsigned(buf []byte, v uint64) []byte {
	if v <= 23 {
		return append(buf, byte(v))
	} else if v <= 0xFF {
		return append(buf, 0x18, byte(v))
	} else if v <= 0xFFFF {
		b := make([]byte, 2)
		binary.BigEndian.PutUint16(b, uint16(v))
		return append(append(buf, 0x19), b...)
	} else if v <= 0xFFFFFFFF {
		b := make([]byte, 4)
		binary.BigEndian.PutUint32(b, uint32(v))
		return append(append(buf, 0x1a), b...)
	}
	b := make([]byte, 8)
	binary.BigEndian.PutUint64(b, v)
	return append(append(buf, 0x1b), b...)
}

// appendCBORBytes encodes a byte string in CBOR format (major type 2).
func appendCBORBytes(buf []byte, data []byte) []byte {
	length := uint64(len(data))
	if length <= 23 {
		buf = append(buf, byte(0x40+length))
	} else if length <= 0xFF {
		buf = append(buf, 0x58, byte(length))
	} else if length <= 0xFFFF {
		b := make([]byte, 2)
		binary.BigEndian.PutUint16(b, uint16(length))
		buf = append(append(buf, 0x59), b...)
	} else {
		b := make([]byte, 4)
		binary.BigEndian.PutUint32(b, uint32(length))
		buf = append(append(buf, 0x5a), b...)
	}
	return append(buf, data...)
}

// decodeCBORUint64 decodes a CBOR header (any major type) and returns the additional value.
// Used for array/map lengths where major type is not 0.
func decodeCBORUint64(data []byte, offset int) (uint64, int, error) {
	if offset >= len(data) {
		return 0, offset, ErrCBORDecode
	}
	additional := data[offset] & 0x1f
	offset++

	if additional <= 23 {
		return uint64(additional), offset, nil
	} else if additional == 24 {
		if offset >= len(data) {
			return 0, offset, ErrCBORDecode
		}
		v := uint64(data[offset])
		return v, offset + 1, nil
	} else if additional == 25 {
		if offset+2 > len(data) {
			return 0, offset, ErrCBORDecode
		}
		v := uint64(binary.BigEndian.Uint16(data[offset : offset+2]))
		return v, offset + 2, nil
	} else if additional == 26 {
		if offset+4 > len(data) {
			return 0, offset, ErrCBORDecode
		}
		v := uint64(binary.BigEndian.Uint32(data[offset : offset+4]))
		return v, offset + 4, nil
	} else if additional == 27 {
		if offset+8 > len(data) {
			return 0, offset, ErrCBORDecode
		}
		v := binary.BigEndian.Uint64(data[offset : offset+8])
		return v, offset + 8, nil
	}
	return 0, offset, ErrCBORDecode
}

// decodeCBORUint decodes a CBOR unsigned integer (major type 0) starting at offset.
// Returns value and new offset.
func decodeCBORUint(data []byte, offset int) (uint64, int, error) {
	if offset >= len(data) {
		return 0, offset, ErrCBORDecode
	}
	majorType := data[offset] >> 5
	if majorType != 0 { // major type 0 = unsigned int
		return 0, offset, ErrCBORDecode
	}
	return decodeCBORUint64(data, offset)
}

// decodeCBORByteString decodes a CBOR byte string starting at offset.
func decodeCBORByteString(data []byte, offset int) ([]byte, int, error) {
	if offset >= len(data) {
		return nil, offset, ErrCBORDecode
	}
	majorType := data[offset] >> 5
	if majorType != 2 { // major type 2 = byte string
		return nil, offset, ErrCBORDecode
	}

	length, newOffset, err := decodeCBORUint64(data, offset)
	if err != nil {
		return nil, offset, err
	}

	end := newOffset + int(length)
	if end > len(data) {
		return nil, newOffset, ErrCBORDecode
	}
	result := make([]byte, length)
	copy(result, data[newOffset:end])
	return result, end, nil
}
